rfd/0057-automatic-user-provisioning.md

```diff
@@ -0,0 +1,123 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic user and sudoers provisioning
+
+## What
+
+Automatically create non-existing users and optionally add them to
+`sudoers` on Teleport nodes. Users will be removed after all sessions
+have logged out.
+
+## Why
+
+Currently, when logging into an SSH node, the user must be
+pre-created. Adding automatic user and `sudoer` provisioning would
+make it so that any Teleport user would be able to login and have the
+account created automatically without manual intervention.
+
+## Details
+
+The following are required for this feature:
+
+- Ability to automatically provision a Linux user if it's not present
+  on the node.
+- Ability to add the provisioned user to existing Linux groups defined
+  in the user traits/role.
+- Ability to add the provisioned user to sudoers.
+- Clean up the provisioned user / sudoers changes upon logout (being
+  careful not to remove pre-existing users).
+
+### Config/Role Changes
+
+Several new fields will need to be added to to the role `options` and
+`allow` sections:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: example
+spec:
+  options:
+    # Controls whether this role supports auto provisioning of users.
+    auto_create_user: true
+    # Controls whether this role allows adding auto provisioned users to sudoers.
+    add_to_sudoers: true
+  allow:
+    # New field listing Linux groups to assign a provisioned user to.
+    # Should support user and identity provider traits like other fields (e.g. "logins")
+    groups: [ubuntu, "{{internal.groups}}", "{{external.xxx}}"]
+```
+
+An individual `ssh_service` can be configured disable auto user
+creation with the below config:
+
+```yaml
+ssh_service:
+    # when disabled, takes precedence over the role setting
+    disable_auto_create_user: true
+```
+
+### User creation
+
+A PAM module can be used in order to avoid running a Teleport process
+as root, and is already recommended in the [Teleport docs](https://goteleport.com/docs/server-access/guides/ssh-pam/#create-local-unix-users-on-login).
+
+This script will be adapted to support multiple Linux distributions
+and their differing methods by which users are created, and can be
+distributed as part of released Teleport packages.
+
+#### User Groups
+
+When a user is created they will be added to the specified groups from
+the role. In addition the user will be added to a special
+`teleport-tmp` group which can be used to indicate that the user was
```

Comment:
Couple of questions:

- How about calling it something like `teleport-system`? `-tmp` suffix just gives impression that this is something debugging related.
- How is this group going to be created on the node?

Reply:
> How is this group going to be created on the node?

I intend to use `groupadd ...` on Teleport daemon startup and create any groups that dont already exist